### PR TITLE
Deve: Budget App Capstone Project (Build and Deploy)

### PR DESCRIPTION
**Pull request by Cosmosha**
**The following requirements were implemented;**

**Splash screen**

- A simple page with the name of your app (yes, you need to choose one), and links to the sign up and log in pages.
Sign up and log in pages

- The user should be able to register in the app with full name, email and password (all mandatory).
- The user can log into the app using email and password.
- If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

**Home page (categories page)**

- When the user logs in, they are presented with the categories page.
- For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.

**Transactions page**

- For a given category, the list of transactions is presented, ordered by the most recent.

**"Add a new category" page**

**"Add a new transaction" page**

**The user fills out a form to create a new transaction with:**
- name (mandatory)
- amount (mandatory)
- categories (mandatory at least one)
- The user click (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
- When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.
- 
**Testing requirements**
Create unit and integration tests for all the most important components of your RoR application.

**Technical requirements**
- You should use Postgres as your database.
- You should use devise for authentication.
- You should validate all user input to make sure that anyone with bad intentions cannot compromise your app.
- You can use a view template engine of your choice (.erb, .slim, .haml).
- The project should be deployed and accessible online.
- Your database schema should reflect the following structure:

### CODE REVIEW

**Review by 1GORDON — COMMENTED** [review]
##
## Hi @Cosmosha ✋

Good job so far!

Hello ✋, you have done well reaching this point of the milestone 🤝 . But there are some issues that you still need to work on to go to the next project but you are almost there!

### This project will be marked as Invalid and won't affect the number of weekly reviews.

- [ ] Kindly ensure your project is using the correct Gitflow. Kindly make sure your feature branch is merging into the development branch instead of the main branch.

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread and tag me `@1GORDON` if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Cosmosha** [comment]
Hello @1GORDON, Please check, I have merged and closed all feature branches I used.

File: app/controllers/entities_controller.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    @entity = @group.entities.new
+  end
+
+  def create
+    @group = Group.find(params[:group_id])
+    @entity = @group.entities.new({ author_id: current_user.id, **entity_params })
+    @entity.save
+
+    redirect_to group_entities_path(@group)
+  end
```
Comment (peterrobert) [inline app/controllers/entities_controller.rb:20]: - [ ] It is always a great practice to create success and failure paths. What if the entity is not created for some reason? You should also handle that scenario.

File: app/controllers/groups_controller.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    @groups = Group.includes(:entities).where(author: current_user)
+  end
+
+  def new; end
+
+  def create
+    current_user.groups.create(group_params)
+    redirect_to groups_path
+  end
+
```
Comment (peterrobert) [inline app/controllers/groups_controller.rb:12]: - [ ] It is always a great practice to create success and failure paths. What if the group is not created for some reason? You should also handle that scenario.

**Review by peterrobert — CHANGES_REQUESTED** [review]
Hi @Cosmosha, 

Very nice project. Great work with your UI. It looks great 👏🏿 . However, there are a few changes to be made kindly check below ⬇️

### Highlights!

[ ] Good feature branch name ✔️
[ ] Descriptive pull request ✔️
[ ]  Awesome work creating simple easy to follow code✔️
[ ] Awesome work handling N+1 problems ✔️

 ## Required Changes ♻️
_Check the comments under the review._

**VIDEO**

- [ ] Great work making the video. However, you missed one of the objectives. It will be great to ensure that the video highlights a piece of code that you are proud of making or that you find interesting.

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread or tag me @peterrobert if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Olamarx — CHANGES_REQUESTED** [review]
Hi @Cosmosha, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted a project, that does not follow gitflow, I will encourage you to make the PR from a feature branch into the deve branch.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

**Comment by Cosmosha** [comment]
Hello @Olamarx , 
I created the pull on the development branch because the project was deployed on render.com from that branch.

**Review by rloterh — COMMENTED** [review]
Hi Cosmos, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted a project, that has incorrect gitflow/ GitHub flow. Kindly note that capstone projects worked on by a single contributor should have Pull Request flow **from feature into development**. You can maintain the deployment on render.com which should not be affected by the gitflow.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

**Pull request closed without merging**